statedb: Allow non-terminated keys

joamaki opened the pull request:
Allow use of indexers that are not fixed-size or which don't terminate the key by checking that the key
length matches exactly with the used search prefix. For unique indexes this is as simple as checking
the key lengths in the iterator and for multi-index we need to extract the length of the secondary key
and compare against that.

E.g. now `struct{A, B string}` with multi-index on `B` will be indexed internally with concatenated key `B + A + <len>`, where `<len>` is 16-bit unsigned big endian int (the KeySet limits key sizes to 16-bits). A `Get` against a multi-index will
return an iterator which will check that the node's key matches with the query, for example `Get("foo")` against
`{"foo", "bar"}` and `{"foob", "ar"}` (assuming no terminating `\0`) will return an iterator which has been
seeked to key "foobar3". The iterator will extract length (3) and compare against search key length (len("foo")) 
and since length matches the object is returned. Next the iterator would advance to "foobar4" which would be
skipped since `4 != len("foo")`. 

In effect this allows statedb to work with indexers that do not terminate the keys, but at the performance cost of
potentially traversing over non-matching tree nodes. The actual length extraction and comparison adds no measurable
time to the iteration (the Get benchmark that was added showed no difference).

This PR additionally extends the tests to cover queries against multi-indexes with overlapping prefixes.

joamaki committed:
statedb: Drop unused unique field from Query

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb: Store index unique info in the index tree entry

To implement special casing for querying unique indexes we need
to know whether the index is unique or not. To avoid extra lookups
store the uniqueness information in the index radix tree.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb: Allow non-terminated keys

Previously statedb required the indexers to either terminate or return fixed-size
keys in order to exactly match with just prefix searching the trees. This is a
fairly large footgun.

This commit adds checks to the iterators to make sure the key matches exactly and is
not a longer key that shares the prefix. For indexers that don't terminate the key,
this may result in wasting bit of time by traversing non-matching nodes.

This was tested by "breaking" index.String by removing termination and validating
that the regression test passes.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb: Add multi-index checks to 29324 regression test

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb: Add benchmark for Get

For checking that the extra key length checks in iteration do not
add any significant cost.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd commented on pkg/statedb/txn.go:
```diff
@@ -353,6 +353,32 @@ func (txn *txn) Delete(meta TableMeta, guardRevision Revision, data any) (any, b
+// properly (e.g. if secondary key is "foo", then we don't want "foobar" to match).
+func encodeMultiIndexKey(primary, secondary []byte) []byte {
+	key := make([]byte, 0, len(secondary)+len(primary)+2)
+	key = append(key, secondary...)
```
`copy`? I'm guessing it's equivalent thanks to compiler optimization, but otherwise this is semantic sugar for appending byte by byte right?

joamaki replied:
Should be fairly equivalent. I find this form easier to reason about.

bimmlerd replied:
ack, no strong preference here - with `copy` you'd have to worry about indices

joamaki committed:
statedb: Add test for query handler

Add a test case for the database seek and iteration part of the
query API handler.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd commented on pkg/statedb/txn.go:
```diff
@@ -300,41 +318,67 @@ func (txn *txn) Delete(meta TableMeta, guardRevision Revision, data any) (any, b
+// This allows looking up from the multi-index with the secondary key by doing a prefix
+// search. The length is used to safe-guard against indexers that don't terminate the key
+// properly (e.g. if secondary key is "foo", then we don't want "foobar" to match).
+func encodeMultiIndexKey(primary, secondary []byte) []byte {
```
applies everywhere: `multi` -> `dual` or `duo` or `secondary`

I think multi is mildly misleading, as these functions don't take more than two indices. (I believe the construction could be extended to support more indices, but I don't see why we'd want to. Uniqueness is guaranteed by the inclusion of the primary index.)

joamaki replied:
This is the key for multi-index (non-unique), hence `MultiIndexKey`, it isn't about the number of indices. And it's kind of incidental that we use the primary index's key here. Just need something unique and deterministic to append that can be derived from the object itself. Could call it `NonUniqueIndexKey`?

EDIT: Ah actually `NonUniqueIndexKey` is better, since we can have unique indexes where objects have multiple keys (they just need to be unique, otherwise they overwrite).

EDIT2: Renamed this and other places to "non-unique"

bimmlerd replied:
ah, I see - works for me!

bimmlerd commented on pkg/statedb/regression_test.go:
```diff
@@ -51,4 +60,17 @@ func Test_Regression_29324(t *testing.T) {
+	// Query on multi-index should only return exact match
+	iter, _ = table.Get(txn, tagIndex.Query("aa"))
+	items = Collect(iter)
+	if assert.Len(t, items, 1, "Get(\"a\") on tags should return one match") {
```
```suggestion
	if assert.Len(t, items, 1, "Get(\"aa\") on tags should return one match") {
```

bimmlerd commented on pkg/statedb/regression_test.go:
```diff
@@ -51,4 +60,17 @@ func Test_Regression_29324(t *testing.T) {
+	// Partial match on prefix should not return anything on multi-index as well
+	iter, _ = table.Get(txn, idIndex.Query("a"))
+	items = Collect(iter)
+	assert.Len(t, items, 0, "Get(\"foob\") should return nothing")
```
```suggestion
	assert.Len(t, items, 0, "Get(\"a\") should return nothing")
```

bimmlerd commented on pkg/statedb/iterator.go:
```diff
@@ -54,6 +55,61 @@ func (it *iterator[Obj]) Next() (obj Obj, revision uint64, ok bool) {
+			return
+		}
+		_, secondary := decodeMultiIndexKey(key)
+		if len(secondary) == len(it.key) {
```
```suggestion
                // Equal length implies equality since as it.key prefixes secondary
		if len(secondary) == len(it.key) {
```

joamaki replied:
Thanks, added comment.

bimmlerd commented on pkg/statedb/txn.go:
```diff
@@ -300,41 +318,67 @@ func (txn *txn) Delete(meta TableMeta, guardRevision Revision, data any) (any, b
+	return binary.BigEndian.AppendUint16(key, uint16(len(secondary)))
+}
+
+func decodeMultiIndexKey(key []byte) (primary []byte, secondary []byte) {
```
This is obviously not resistant against tampering. That's fine, I believe, as long as we don't introduce a persistence layer which takes these lengths at face value. I don't know how to add a reminder about this to ourselves in the future :sweat_smile:

joamaki replied:
If we're concerned about tampering this is probably one of the more minor concerns :D. Besides I don't really see a reason to persist the indexes. I would just persist the object itself and rebuild indexes on load.

bimmlerd replied:
fair point :grin: let's not worry about it here.

bimmlerd reviewed (commented):
approach LGTM, nits

bimmlerd reviewed (approved):
Is the intention to then generally drop termination etc from the indexing functions or is this purely a safeguard against mistakes in those functions?
